geth/api/api.go

```diff
@@ -52,65 +53,88 @@ func (api *StatusAPI) TxQueueManager() *transactions.Manager {
 
 // StartNode start Status node, fails if node is already started
 func (api *StatusAPI) StartNode(config *params.NodeConfig) error {
-	nodeStarted, err := api.b.StartNode(config)
-	if err != nil {
-		return err
-	}
-	<-nodeStarted
-	return nil
+	return api.b.StartNode(config)
 }
 
 // StartNodeAsync start Status node, fails if node is already started
 // Returns immediately w/o waiting for node to start (see node.ready)
 func (api *StatusAPI) StartNodeAsync(config *params.NodeConfig) (<-chan struct{}, error) {
-	return api.b.StartNode(config)
+	resp := make(chan struct{})
+	if api.b.IsNodeRunning() {
+		close(resp)
+		return resp, nil
+	}
+	go func() {
+		api.StartNode(config) // nolint: errcheck
```

Comment:
I think we should report an error in an async interface as well. Generally, we should come up with a generic way of creating an async counterpart of a sync method that can be interrupted early (if possible and make sense) and return an error.

Maybe something like this:
```go
func (api *StatusAPI) StartNodeAsync(ctx context.Context, config *params.NodeConfig) (<-chan error)
```

In this case, `ctx` is not really helpful though. The returned channel can be used like this:
```go
select {
case err := <-err:
  if err != nil {
    // error occured
  }

  // success
}
```

However, I am not really sure we should use a single channel to report on separate issues: (1) that the operation is done and (2) an error occurred. Also, the drawback is that for the reader, the output will be different: https://play.golang.org/p/OFzpNn6hH4X. Also, if there will be more than one reader, only one will be able to read the error.

Reply:
Is it possible to hide this channel behind some other type to offer more readable code?
```go
err := api.StartNodeAsync().Wait()
```

Reply:
We will end up with some sort of Future/Promise implementation. It would be nice if we could only use Go primitives. 

I can think of:
1. Return an empty channel which only tells when the operation finished:
```go
done := api.StartNodeAsync()
<-done
// check error
if api.Err(); err != nil {
  // do something
}
// but how the caller can be sure `api.Err()` returns an error of that particular call?
```

2. Return an object similar to what you described:
```go
future := api.StartNodeAsync()
<-future.Done()
if err := future.Err(); err != nil {
  // do something
}
```
However, I haven't really seen this approach often which makes me think it's possible to handle it better. Maybe we shouldn't equip `StatusBackend` with async methods as it's not the last part of the call chain. Maybe we should remove signals from `NodeManager` and `Node` and use them to actually signalize errors. We would end up with something like this:
```go
// StatusBackend has only well-defined, sync methods like StartNode(). 

// There is StatusBackendAsyncSignal adapter and we agree that the caller expects errors only via signals.
type StatusBackendAsyncSignal struct {
  *StatusBackend
}

func (b *StatusBackendAsyncSignal) StartNodeAsync() <-chan struct{} {
  // not sure done is even needed but may be useful for testing
  done := make(chan struct{})
  func () {
    defer close(done)

    err := b.StartNode()
    if err != nil {
      signal.Send({ ... })
    }
  }()

  return done
}
```

Reply:
previously if another node was already running error returned to a caller immediately, even before spawning goroutine. i changed it to reflect intent better

- stop the node  - node is already stopped, no need for an error
- start the node - node is already running, no error
- restart the node - failure returned, intent is not clear :)

but i guees it is not really necessary, i will use idea with error chanel

Reply:
i like the idea with error channel, 
and at the moment we always have one consumer, so i would just use proposed approach and avoid unnecessary complexity

Reply:
I will leave signals in Backend for now though, as there are 2 places where I can't move them to API